### PR TITLE
Fix bug in s3Cluster schema inference

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug in s3Cluster schema inference that let to the fact that not all data was read in the select from s3Cluster.
The bug appeared in https://github.com/ClickHouse/ClickHouse/pull/35544.

cc: @nikitamikhaylov

### CODE REVIEW

File: src/Storages/StorageS3.cpp [hunk 1]
```diff
@@ -182,24 +182,21 @@ String StorageS3Source::DisclosedGlobIterator::next()
...
     {
-        std::lock_guard lock(mutex);
-        if (keys_iter == keys.end())
+        size_t current_index = index.fetch_add(1);
+        if (current_index >= keys.size())
             return "";
-        auto key = *keys_iter;
-        ++keys_iter;
-        return key;
+        return keys[current_index];
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:194]: These changes are not related to bug, just decided to use atomic instead of mutex here

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -182,24 +182,21 @@ String StorageS3Source::DisclosedGlobIterator::next()
...
-    explicit Impl(const std::vector<String> & keys_) : keys(keys_), keys_iter(keys.begin())
+    explicit Impl(const std::vector<String> & keys_) : keys(keys_)
     {
     }
 
     String next()
     {
-        std::lock_guard lock(mutex);
-        if (keys_iter == keys.end())
+        size_t current_index = index.fetch_add(1);
```
Comment (antonio2368) [inline src/Storages/StorageS3.cpp:191]: I think that this could be `relaxed` as the threads only modify `index` and only here.

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Comment by nikitamikhaylov** [comment]
@Avogar 😥
